imap/mboxlist.c

```diff
@@ -5699,26 +5714,139 @@ EXPORTED int mboxlist_upgrade(int *upgraded)
     return r;
 }
 
+static int subsdb_needs_upgrade(const char *userid,
+                                const char *subsfname,
+                                struct db *subs,
+                                int *curr_version)
+{
+    const char *key = SUBDB_VERSION_KEY;
+    size_t keylen = strlen(SUBDB_VERSION_KEY);
+    const char *data = NULL;
+    size_t datalen = 0;
+    char *tmp, *endptr;
+    int version, r;
+
+    r = cyrusdb_fetch(subs, key, keylen, &data, &datalen, NULL);
+    switch (r) {
+    case CYRUSDB_NOTFOUND:
+        version = 0;
+        break;
+    case CYRUSDB_OK:
+        /* need a copy that's definitely null-terminated for strtol */
+        tmp = xstrndup(data, datalen);
+        version = strtol(tmp, &endptr, 10);
```

Comment:
nit: Should probably check errno for overflow/underflow?

Reply:
Hmm.  I think if that happens then whatever remaining digits it couldn't parse will wind up in *endptr, and we'll hit the error check below.  But I'm rereading the man page now, and it's not clear what happens with endptr on overflow.  So yeah I'd better tighten that up